# 01. Import libraries

In [1]:
# Importing libraries (pandas, numpy, os)
import pandas as pd
import numpy as np
import os

# 02. Import Data

In [7]:
# Folder path: 
path = r"C:\Users\Windows\Instacart Basket Analysis"

In [8]:
# Import products.csv: 
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

NameError: name 'pd' is not defined

In [5]:
# Import orders_wrangled.csv: 
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'))

# 03. Consistency Check

### Mixed Data Types

In [6]:
# Create test dataframe:
df_test = pd.DataFrame()

In [7]:
# Create a mixed column type:
df_test['mix'] = ['a', 'b', 1, True]

In [8]:
# Check for mixed data types:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col, ' mixed')
  else: print(col, ' consistent')

mix  mixed


In [9]:
# Changing 'mix' to string type:
df_test['mix'] = df_test['mix'].astype('str')

In [10]:
# Re-check for mixed data types:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col, ' mixed')
  else: print(col, ' consistent')

mix  consistent


### Missing Values

In [11]:
# Count the missing vailes in each col:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
# Missing Values list: 
df_nanprods = df_prods[df_prods.product_name.isnull() == True]

In [13]:
# Create df without nan values:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [14]:
df_prods_clean.shape

(49677, 5)

### Duplicates

In [15]:
# Create duplicates df:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [16]:
# Creat dropped dups list 
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [17]:
df_prods_clean_no_dups.shape

(49672, 5)

### Export Changes

In [18]:
# Export changes df:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'df_prods_cleaned.csv'))

# 4.5: Data Consistency Checks

In [19]:
# Step 2. Run the df.describe() function on your df_ords dataframe:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

Question 2: Min value = 1.000000, Max value 99999.000000, perhaps this could been a error. 

In [20]:
df_prods[df_prods.prices == 99999]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

In [6]:
# Step 3. Check for mixed-type data in your df_ords dataframe:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
        print (col, ' mixed')
  else: print(col, ' consistent')

NameError: name 'df_ords' is not defined

In [62]:
# Step 4. If you find mixed-type data, fix it:

The column in question should contain observations of a single data type:
No mixed-type data found. 

In [22]:
# Step 5. Run a check for missing values in your df_ords dataframe:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
hour_of_day_ordered            0
days_since_prior_order    206209
dtype: int64

In [5]:
df_ords.head(20)

NameError: name 'df_ords' is not defined

In [23]:
df_ords['user_id'] = df_ords['user_id'].astype('int64')
df_ords.user_id.max()

206209

In [24]:
# create a dataframe for the missing values
df_ords_nan = df_ords [df_ords ['days_since_prior_order'].isnull () == True]

In [54]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

The reason for the missing value is perhaps because the majority of customers are first-time clients who have never placed an order before. In addition, there won't be a value indicating how long it has been since a prior order. The maximum value in the column 'user_id' is 206209, which corresponds with the number of missing values in the column 'days_since_prior_order" that represents one for each new customer.

In [25]:
# Step 6. Address the missing values using an appropriate method:

I’ve chosen to make no changes. This is because I believe that it isn’t reasonable to assign values or remove or filter the data. Statistics may still be analyzed with the NaN values present. A value of "1’ in the column 'order_number' can represent new customers.

In [26]:
# Step 7. Run a check for duplicate values in your df_ords data: 
df_ords_dups = df_ords[df_ords.duplicated()]

In [27]:
df_ords_dups.size

0

No duplicates found

In [28]:
# Step 8. Address the duplicates using an appropriate method:

I’ve chosen to make no changes, following step 6 action. 

In [4]:
# Step 9. Export changes: 
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'df_ords_cleaned1.csv'))

NameError: name 'df_ords' is not defined